Components: Add ValidateControls components

gigitux opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->
Related PR: https://github.com/Automattic/wp-calypso/pull/100771

## What?

This PR adds the `ValidateControls` components as private APIs. These components wrap the WordPress components adding the validation controls too.

This PR imports https://github.com/Automattic/wp-calypso/pull/100771 changes with small changes:

- fixed ESLint rules
- updated the CSS selectors replacing `a8c` with `components`
- restructure folders following the Gutenberg patterns.


The first commit contains the raw copy from Calypso repo (https://github.com/WordPress/gutenberg/pull/70620/commits/6ac9b6f4c2c53fa8355489c24882c2a5f02f480a)

 

<!-- In a few words, what is the PR actually doing? -->

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

We’re exploring how to introduce validation in the **DataViews** library (https://github.com/Automattic/wp-calypso/pull/104145 to be migrated to Gutenberg soon). As part of this effort, we need components that support validation to build and test the integration effectively.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
1. Run `npm run storybook:dev`.
2. Smoke test the `Validated Form Controls` stories.



### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<!-- If you would like to upload screenshots, feel free to use the table below when it is useful to show the difference between before and after the change. -->

|Before|After|
|-|-|
|<!-- Before screenshot here -->|<!-- After screenshot here -->|

gigitux committed:
import components from calypso

gigitux committed:
fix folder structure

gigitux committed:
export as private API

gigitux committed:
fix build

gigitux committed:
rename stories file

gigitux committed:
fix storybook

gigitux committed:
update stories title

gigitux committed:
add changelog

mirka commented on packages/components/src/validated-form-controls/index.ts:
```diff
@@ -0,0 +1,2 @@
+export * from './components';
+export * from './control-with-error';
```
We're considering this a stopgap until we have a more integrated set of components. Would it be possible to export only the components that are absolutely needed for the product requirements at hand? I'm assuming this is specifically for binding validated components to DataForm — do you know which specific form components are needed?

Related: It would be great if you could note this in the "Why" section of the PR description so everyone can see why we need these components now.

gigitux replied:
Good point! Honestly, I’m not sure yet. If you’re on board, I’d suggest not exporting any components, not even as a private API, until we actually need them. Would that work for you?

mirka replied:
Sounds good 👍 We can add any private exports as needed.

mirka commented on packages/components/CHANGELOG.md:
```diff
@@ -2,6 +2,11 @@
+
+### Enhancement
+
+-   Add new validated form controls: `ValidatedCheckboxControl`, `ValidatedInputControl`, `ValidatedNumberControl`, `ValidatedRadioControl`, `ValidatedRangeControl`, `ValidatedSelectControl`, `ValidatedTextControl`, `ValidatedTextareaControl`, `ValidatedToggleControl`, `ValidatedCustomSelectControl`, `ValidatedComboboxControl`, `ValidatedToggleGroupControl`, and `ControlWithError` for enhanced fo ...
```
We might want to note that these components are private, especially since there's a good chance we won't ever be exposing them publicly.

mirka replied:
We can match this [similar entry](https://github.com/WordPress/gutenberg/pull/70681/files#diff-b5aa808176801bfde2277ac886bbbdf0863de109950303cdb89e2d33ebe16a28).

gigitux replied:
I updated the changelog with 31c0ab2782b3b9904583d582df127d7cfa6a9e4d. You raised a [good point](https://github.com/WordPress/gutenberg/pull/70620#discussion_r2198064492). We need to update the changelog if we decide not to export any component in this PR.

mirka commented on packages/components/src/validated-form-controls/components/stories/overview.mdx:
```diff
@@ -0,0 +1,52 @@
+
+## Status: Beta
+
+We are still gathering feedback and iterating. Please get in touch with Team Calypso if you are interested in trying them out in your project.
```
```suggestion
We are still gathering feedback and iterating. Please get in touch with `@WordPress/gutenberg-components`  if you are interested in trying them out in the Gutenberg app.
```

gigitux replied:
I missed this! Thanks! Fixed with b80006d9a1c9d4b4ae371f6ebad1f2b234276edf

mirka commented on packages/components/src/validated-form-controls/components/stories/checkbox-control.story.tsx:
```diff
@@ -0,0 +1,56 @@
+import { formDecorator } from './story-utils';
+
+const meta: Meta< typeof ValidatedCheckboxControl > = {
+	title: 'Components (Private)/Validated Form Controls/ValidatedCheckboxControl',
```
```suggestion
	title: 'Components (Experimental)/Validated Form Controls/ValidatedCheckboxControl',
```

How about we move these all to the experimental components section, and add a "private" tag like the other locked components?

https://github.com/WordPress/gutenberg/blob/8189c9304bd6749f8ef43aa446e27e34eed245e5/packages/components/src/menu/stories/index.story.tsx#L58

gigitux replied:
Addressed with 35c8eb47f49eca8d9b9ca2363a765dd19edd8e33

mirka commented on packages/components/src/validated-form-controls/style.scss:
```diff
@@ -0,0 +1,75 @@
+	align-items: flex-start;
+	gap: 4px;
+	margin: 8px 0 0;
+	// font-family: typography.$sans;
```
There's currently a quirk with the `@use` path resolution in this repo, but let's continue using the explicit syntax. There's also a variable we can use for the font-family here.

```diff
diff --git a/packages/components/src/validated-form-controls/style.scss b/packages/components/src/validated-form-controls/style.scss
index 93d0a425ec..3e440d2f1c 100644
--- a/packages/components/src/validated-form-controls/style.scss
+++ b/packages/components/src/validated-form-controls/style.scss
@@ -1,23 +1,26 @@
+@use "colors";
+@use "variables";
+
 .components-validated-control {
 	// For components based on InputBase
 	&:has(:is(input, select):user-invalid)
 	.components-input-control__backdrop {
-		--wp-components-color-accent: $alert-red;
-		border-color: $alert-red;
+		--wp-components-color-accent: colors.$alert-red;
+		border-color: colors.$alert-red;
 	}
 
 	// For TextControl, TextareaControl
 	:is(textarea, input[type="text"]):user-invalid {
-		--wp-admin-theme-color: $alert-red;
-		--wp-components-color-accent: $alert-red;
-		border-color: $alert-red;
+		--wp-admin-theme-color: colors.$alert-red;
+		--wp-components-color-accent: colors.$alert-red;
+		border-color: colors.$alert-red;
 	}
 
 	// For ComboboxControl
 	.components-combobox-control__suggestions-container:has(
 	input:user-invalid
 ):not(:has([aria-expanded="true"])) {
-		border-color: $alert-red;
+		border-color: colors.$alert-red;
 	}
 }
 
@@ -26,13 +29,13 @@
 
 	// For CustomSelectControl
 	&:has(select:user-invalid) .components-input-control__backdrop {
-		--wp-components-color-accent: $alert-red;
-		border-color: $alert-red;
+		--wp-components-color-accent: colors.$alert-red;
+		border-color: colors.$alert-red;
 	}
 
 	// For ToggleGroupControl
 	&:has(input[type="radio"]:invalid) {
-		--wp-components-color-accent: $alert-red;
+		--wp-components-color-accent: colors.$alert-red;
 	}
 }
 
@@ -50,10 +53,10 @@
 	align-items: flex-start;
 	gap: 4px;
 	margin: 8px 0 0;
-	// font-family: typography.$sans;
+	font-family: variables.$font-family-body;
 	font-size: 0.75rem;
 	line-height: 16px; // matches the icon size
-	color: $alert-red;
+	color: colors.$alert-red;
 	animation:
 		components-validated-control__error-jump 0.2s
 		cubic-bezier(0.68, -0.55, 0.27, 1.55);
```

gigitux replied:
Thanks! I set the font-family with e83718c032e6d0cfb5c71fe2b8f74f74e5f1794a

mirka commented on packages/components/src/validated-form-controls/components/checkbox-control.tsx:
```diff
@@ -0,0 +1,61 @@
+		onChange,
+		markWhenOptional,
+		...restProps
+	}: Omit< CheckboxControlProps, '__nextHasNoMarginBottom' > &
```
`CheckboxControlProps` is actually just a subset of `React.ComponentProps< typeof CheckboxControl >`, and we need the latter set here. (Same with all the other components)

```suggestion
	}: Omit< React.ComponentProps< typeof CheckboxControl >, '__nextHasNoMarginBottom' > &
```

gigitux replied:
Addressed with c571df83706b256094aae8e0a9bfec444d507e4e

ciampo commented:
Apologies @mirka I clicked on the "request review" button by mistake while tabbing through the UI 😅

gigitux committed:
Update packages/components/src/validated-form-controls/components/stories/overview.mdx

Co-authored-by: Lena Morita <lena@jaguchi.com>

gigitux committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into fix/add-controls

gigitux committed:
use variable for font-family

gigitux committed:
add status-private tag to stories

gigitux committed:
update changelog

gigitux committed:
improve props type definition

gigitux committed:
Merge branch 'fix/add-controls' of github.com:WordPress/gutenberg into fix/add-controls

mirka reviewed (approved):
Ready to merge once the exports are removed and the readme is updated 🚀

gigitux committed:
remove export validated form controls

gigitux commented:
> Ready to merge once the exports are removed and the readme is updated 🚀

Thanks for the review! I removed the export and updated the changelog with 714adc368fc03a8e53c3d6bf4073e87a81af8b15.

gigitux commented:
@mirka Let me know if you prefer to take another look at my last commits, or I can merge the PR. Thanks! 🙇
